File: QuickSort.java
public class QuickSort {
    int[] array;

    public QuickSort(int[] array) {
        this.array = array;
    }
    public void quickSort() {
        quickSort(0, array.length - 1);
    }
    public void quickSort(int start, int end) {
        if (start >= end) {
            return;
        }
        int pivot = partition(start, end);
        quickSort(0, pivot - 1);
        quickSort(pivot + 1, end);
    }
    private int partition(int start, int end) {
        int pIndex = start;
        int pivot = array[end];
        for (int i = start; i < end; i++) {
            if (array[i] <= pivot) {
                swap(i, pIndex);
                pIndex++;
            }
        }
        swap(pIndex, end);
        return pIndex;
    }
    private void swap(int location1, int location2) {
        int temp = array[location1];
        array[location1] = array[location2];
        array[location2] = temp;
    }
}

File: quicksort.py
class quicksort:

    def __init__(self, array):
        self.array = array
    def partition(self, left, right):
        pivot = self.array[right]
        boundarySwap = left
        for i in range(left, right):
            if self.array[i] <= pivot:
                temp = self.array[boundarySwap]
                self.array[boundarySwap] = self.array[i]
                self.array[i] = temp
                boundarySwap += 1
        temp = self.array[boundarySwap]
        self.array[boundarySwap] = self.array[right]
        self.array[right] = temp
        return boundarySwap

    def quicksortHelper(self, left, right):
        if (left < right):
            pivotIndex = self.partition(left, right)
            self.quicksortHelper(left, pivotIndex - 1)
            self.quicksortHelper(pivotIndex + 1, right)
    def quicksort(self):
        self.quicksortHelper(0, len(self.array) - 1)

    def print_array(self):
        for i in self.array:
            print(i, end =" ")
